src/main/starlark/builtins_bzl/common/cc/compile/compile.bzl

```diff
@@ -436,6 +493,602 @@ def _should_provide_header_modules(
+            copts = copts,
+            copts_filter = copts_filter,
+            cpp_configuration = cpp_configuration,
+            cxxopts = cxxopts,
+            feature_configuration = feature_configuration,
+            label = label,
+            common_toolchain_variables = common_compile_build_variables,
+            language = language,
+            native_cc_semantics = native_cc_semantics,
+            source_artifact = source_artifact,
+            source_label = source_label,
+            use_pic = use_pic,
+            ddi_file = ddi_file,
+            ddi_output_name = ddi_output_name,
+        )
+        source_to_ddi_file_map[source_artifact] = ddi_file
+
+    _create_aggregate_ddi_action(
+        actions = actions,
+        cc_toolchain = cc_toolchain,
+        ddi_files = source_to_ddi_file_map.values(),
+        direct_module_files = direct_module_files,
+        transitive_modules_info_files = cc_compilation_context.modules_info_files(use_pic = use_pic),
+        modules_info_file = modules_info_file,
+    )
+    compiled_basenames = set()
+    all_module_files = depset(direct_module_files, transitive = [cc_compilation_context.module_files(use_pic = use_pic)])
+    for cpp_source in module_interfaces_sources.values():
+        source_artifact = cpp_source.file
+        output_name = output_name_map[source_artifact]
+        source_label = cpp_source.label
+        bitcode_output = feature_configuration.is_enabled("thin_lto") and (("." + source_artifact.extension) in LTO_SOURCE_EXTENSIONS)
+        module_file = source_to_module_file_map[source_artifact]
+        if use_pic:
+            output_name_base = cc_internal.get_artifact_name_for_category(
+                cc_toolchain = cc_toolchain,
+                category = artifact_category.PIC_FILE,
+                output_name = output_name,
+            )
+            outputs.add_pic_cpp20_module_file(module_file)
+        else:
+            output_name_base = output_name
+            outputs.add_cpp20_module_file(module_file)
+
+        modmap_file = _get_compile_output_file(
+            action_construction_context,
+            label,
+            output_name = cc_internal.get_artifact_name_for_category(
+                cc_toolchain = cc_toolchain,
+                category = artifact_category.CPP_MODULES_MODMAP,
+                output_name = output_name_base,
+            ),
+        )
+        modmap_input_file = _get_compile_output_file(
+            action_construction_context,
+            label,
+            output_name = cc_internal.get_artifact_name_for_category(
+                cc_toolchain = cc_toolchain,
+                category = artifact_category.CPP_MODULES_MODMAP_INPUT,
+                output_name = output_name_base,
+            ),
+        )
+        _create_gen_modmap_action(
+            actions = actions,
+            cc_toolchain = cc_toolchain,
+            ddi_file = source_to_ddi_file_map.get(source_artifact),
```

Comment:
Do we ever expect `source_artifact not in source_to_ddi_file_map`? If so, please add a comment explaining what happens in that case, if not, use `[...]` instead of `get` to make this an error.